File: 89.gray-code.js
/*
 * @lc app=leetcode id=89 lang=javascript
 *
 * [89] Gray Code
 */

// @lc code=start
/**
 * @param {number} n
 * @return {number[]}
 */
var grayCode = function (n) {
  const res = [0];
  for (let i = 0; i < n; ++i) {
    for (let j = res.length - 1; j >= 0; --j) {
      res.push(res[j] | (1 << i));
    }
  }
  return res;
};
// @lc code=end

File: 89.gray-code.py
#
# @lc app=leetcode id=89 lang=python3
#
# [89] Gray Code
#

# Time:  O(2^n)
# Space: O(1)

from typing import List


# @lc code=start
class Solution:
    def grayCode(self, n: int) -> List[int]:
        res = [0]
        for i in range(n):
            size = len(res)
            for j in range(size - 1, -1, -1):
                res.append(res[j] | (1 << i))
        return res

        res = []
        for i in range(2 ** n):
            res.append((i >> 1) ^ i)
        return res


# @lc code=end
